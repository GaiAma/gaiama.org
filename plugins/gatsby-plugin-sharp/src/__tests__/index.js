const path = require(`path`)

const {
  base64,
  responsiveSizes,
  resolutions,
  queueImageResizing,
} = require(`../`)

describe(`gatsby-plugin-sharp`, () => {
  const args = {
    duotone: false,
    grayscale: false,
    rotate: false,
  }
  const absolutePath = path.join(__dirname, `images/test.png`)
  const file = getFileObject(absolutePath)

  describe(`queueImageResizing`, () => {
    it(`should round height when auto-calculated`, () => {
      // Resize 144-density.png (281x136) with a 3px width
      const result = queueImageResizing({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args: { width: 3 },
      })

      // Width should be: w = (3 * 136) / 281 = 1.451957295
      // We expect value to be rounded to 1
      expect(result.height).toBe(1)
    })
  })

  describe(`responsiveSizes`, () => {
    it(`includes responsive image properties, e.g. sizes, srcset, etc.`, async () => {
      const result = await responsiveSizes({ file })

      expect(result).toMatchSnapshot()
    })

    it(`adds pathPrefix if defined`, async () => {
      const pathPrefix = `/blog`
      const result = await responsiveSizes({
        file,
        args: {
          pathPrefix,
        },
      })

      expect(result.src.indexOf(pathPrefix)).toBe(0)
      expect(result.srcSet.indexOf(pathPrefix)).toBe(0)
    })

    it(`keeps original file name`, async () => {
      const result = await responsiveSizes({
        file,
      })

      expect(result.src.indexOf(file.name)).toBe(8)
      expect(result.srcSet.indexOf(file.name)).toBe(8)
    })

    it(`accounts for pixel density`, async () => {
      const result = await responsiveSizes({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args: {
          sizeByPixelDensity: true,
        },
      })

      expect(result).toMatchSnapshot()
    })

    it(`can optionally ignore pixel density`, async () => {
      const result = await responsiveSizes({
        file: getFileObject(path.join(__dirname, `images/144-density.png`)),
        args: {
          sizeByPixelDensity: false,
        },
      })

      expect(result).toMatchSnapshot()
    })

    it(`does not change the arguments object it is given`, async () => {
      const args = { maxWidth: 400 }
      await responsiveSizes({
        file,
        args,
      })

      expect(args).toEqual({ maxWidth: 400 })
    })
  })

  describe(`resolutions`, () => {
    console.warn = jest.fn()

    beforeEach(() => {
      console.warn.mockClear()
    })

    afterAll(() => {
      console.warn.mockClear()
    })

    it(`does not warn when the requested width is equal to the image width`, async () => {
      const args = { width: 1 }

      const result = await resolutions({
        file,
        args,
      })

      expect(result.width).toEqual(1)
      expect(console.warn).toHaveBeenCalledTimes(0)
    })

    it(`warns when the requested width is greater than the image width`, async () => {
      const args = { width: 2 }

      const result = await resolutions({
        file,
        args,
      })

      expect(result.width).toEqual(1)
      expect(console.warn).toHaveBeenCalledTimes(1)
    })
  })

  describe(`base64`, () => {
    it(`converts image to base64`, async () => {
      const result = await base64({
        file,
        args,
      })

      expect(result).toMatchSnapshot()
    })
  })

  describe(`image quirks`, () => {
    /**
     * as of https://github.com/nodeca/probe-image-size/issues/20
     */
    it(`handles padding bytes correctly`, async () => {
      const result = await responsiveSizes({
        file: getFileObject(path.join(__dirname, `images/padding-bytes.jpg`)),
        args: {
          sizeByPixelDensity: false,
        },
      })

      expect(result).toMatchSnapshot()
    })
  })
})

function getFileObject(absolutePath) {
  return {
    id: `${absolutePath} absPath of file`,
    name: `test`,
    absolutePath,
    extension: `png`,
    internal: {
      contentDigest: `1234`,
    },
  }
}
