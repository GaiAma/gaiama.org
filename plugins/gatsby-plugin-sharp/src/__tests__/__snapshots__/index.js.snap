// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin-sharp base64 converts image to base64 1`] = `
Object {
  "aspectRatio": 1,
  "height": 20,
  "originalName": undefined,
  "src": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAIAAAAC64paAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAGElEQVQ4y2P4TwFgGNU8qnlU86jmgdUMAOBEq42KgAyMAAAAAElFTkSuQmCC",
  "width": 20,
}
`;

exports[`gatsby-plugin-sharp image quirks handles padding bytes correctly 1`] = `
Object {
  "aspectRatio": 0.746,
  "base64": "data:image/jpeg;base64,/9j/2wBDABALDA4MChAODQ4SERATGCgaGBYWGDEjJR0oOjM9PDkzODdASFxOQERXRTc4UG1RV19iZ2hnPk1xeXBkeFxlZ2P/2wBDARESEhgVGC8aGi9jQjhCY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2P/wgARCAAbABQDASIAAhEBAxEB/8QAGQAAAgMBAAAAAAAAAAAAAAAAAAQCBQYD/8QAFwEAAwEAAAAAAAAAAAAAAAAAAAECA//aAAwDAQACEAMQAAABUSuk86QHBj3a3ijKGjLn/8QAGxAAAgMBAQEAAAAAAAAAAAAAAQIAAxETEiH/2gAIAQEAAQUCuwHPnmWOItZ6N0JVH19jPhhUNONc/8QAFhEBAQEAAAAAAAAAAAAAAAAAEQAQ/9oACAEDAQE/ATGb/8QAFhEBAQEAAAAAAAAAAAAAAAAAAQAC/9oACAECAQE/AXcLOYv/xAAhEAABAgQHAAAAAAAAAAAAAAABABEQEiGBAgMjMTJBUf/aAAgBAQAGPwLTMw7LR2sFNmCnicINhN1WNQ64Bf/EAB4QAQACAQQDAAAAAAAAAAAAAAEAEUEQITFxYYGR/9oACAEBAAE/IReQzcsHuC4Z5isStliiCFTzTagrqP7uiioWIIqhvWhVF2JRx8J//9oADAMBAAIAAwAAABCr4D7/xAAXEQEBAQEAAAAAAAAAAAAAAAABABEh/9oACAEDAQE/EDBspJeZI3l//8QAGREBAQADAQAAAAAAAAAAAAAAAQARITGR/9oACAECAQE/ENCyDOvbLxgBf//EAB4QAQACAgMAAwAAAAAAAAAAAAEAESExQVFhcYGh/9oACAEBAAE/EG+DAkcBTM7w7OiW7Cj03XkaQBDCd1QahgEVpTWh8wTYmNcPIn5TVVfaCcEGxKr5ghUwtXP7CUA3QMVoEj//2Q==",
  "density": 72,
  "originalImg": "/static/test-1234-ae944.png",
  "originalName": undefined,
  "presentationHeight": 1000,
  "presentationWidth": 746,
  "sizes": "(max-width: 746px) 100vw, 746px",
  "src": "/static/test-1234-ae944.png",
  "srcSet": "/static/test-1234-8d5df.png 200w,
/static/test-1234-2c1d9.png 400w,
/static/test-1234-ae944.png 746w",
}
`;

exports[`gatsby-plugin-sharp responsiveSizes accounts for pixel density 1`] = `
Object {
  "aspectRatio": 2.0661764705882355,
  "base64": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAKCAYAAAC0VX7mAAAACXBIWXMAABYlAAAWJQFJUiTwAAABAklEQVQoz61R7WrCQBDM+7+Cv1qhlIogoUWMoH8KrfgOLVXRKDV4d7kkl6/pbsxpqBIo9mBuZ7N3k51bB7TKsrwZVsexyS2rKerwFhYFVI1DnkMQOHIuCaLB+ayuwXlMaApXgvMoxjTU8MIQE4rMx8SnNffUkU/qM/bbiPAeRfgw5n8tB+QgO1kmUvwROdm0kYHmUNoe+NoU2wZTdVjQH7Isg9YRpFIXor8vpGmKjb/FQQgslit8fi3wvQ/OHXLhxRvDHY7Qd5+x3e2udmNzTUN4fZvhoddH566L+8cn9AZuVcvpLR3e9kEAISWEkDAmbRXkGCcJ1r4PRRNPEkPu9Kn2Azs0CqJYU3JKAAAAAElFTkSuQmCC",
  "density": 144,
  "originalImg": "/static/test-1234-85b96.png",
  "originalName": undefined,
  "presentationHeight": 68,
  "presentationWidth": 141,
  "sizes": "(max-width: 141px) 100vw, 141px",
  "src": "/static/test-1234-85b96.png",
  "srcSet": "/static/test-1234-7c9f5.png 200w,
/static/test-1234-85b96.png 281w",
}
`;

exports[`gatsby-plugin-sharp responsiveSizes can optionally ignore pixel density 1`] = `
Object {
  "aspectRatio": 2.0661764705882355,
  "base64": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAKCAYAAAC0VX7mAAAACXBIWXMAABYlAAAWJQFJUiTwAAABAklEQVQoz61R7WrCQBDM+7+Cv1qhlIogoUWMoH8KrfgOLVXRKDV4d7kkl6/pbsxpqBIo9mBuZ7N3k51bB7TKsrwZVsexyS2rKerwFhYFVI1DnkMQOHIuCaLB+ayuwXlMaApXgvMoxjTU8MIQE4rMx8SnNffUkU/qM/bbiPAeRfgw5n8tB+QgO1kmUvwROdm0kYHmUNoe+NoU2wZTdVjQH7Isg9YRpFIXor8vpGmKjb/FQQgslit8fi3wvQ/OHXLhxRvDHY7Qd5+x3e2udmNzTUN4fZvhoddH566L+8cn9AZuVcvpLR3e9kEAISWEkDAmbRXkGCcJ1r4PRRNPEkPu9Kn2Azs0CqJYU3JKAAAAAElFTkSuQmCC",
  "density": 144,
  "originalImg": "/static/test-1234-24ad8.png",
  "originalName": undefined,
  "presentationHeight": 136,
  "presentationWidth": 281,
  "sizes": "(max-width: 281px) 100vw, 281px",
  "src": "/static/test-1234-24ad8.png",
  "srcSet": "/static/test-1234-06d78.png 200w,
/static/test-1234-24ad8.png 281w",
}
`;

exports[`gatsby-plugin-sharp responsiveSizes includes responsive image properties, e.g. sizes, srcset, etc. 1`] = `
Object {
  "aspectRatio": 1,
  "base64": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAIAAAAC64paAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAGElEQVQ4y2P4TwFgGNU8qnlU86jmgdUMAOBEq42KgAyMAAAAAElFTkSuQmCC",
  "density": 72,
  "originalImg": "/static/test-1234-18c2a.png",
  "originalName": undefined,
  "presentationHeight": 1,
  "presentationWidth": 1,
  "sizes": "(max-width: 1px) 100vw, 1px",
  "src": "/static/test-1234-18c2a.png",
  "srcSet": "/static/test-1234-18c2a.png 1w",
}
`;
